Colonel Mustard in the Lounge with the Lead Pipe (killed mapBlockIndex outside of validation)

TheBlueMatt opened the pull request:
Replaces #10692 after #11041 was merged. Stops exporting mapBlockIndex in validation.h entirely, also making it const inside of validation.cpp outside of CChainState. Also makes chainActive const everywhere outside of CChainState and, together, removes all non-const CBlockIndex*s outside of validation/(mostly CChainState).

Sorry, no lyrics this time.

TheBlueMatt committed:
Add a const_cast'ed reference to chainActive in miner_tests

TheBlueMatt committed:
Make CChain only return const CBlockIndex* in const functions

TheBlueMatt committed:
Make CBlockIndex* arguments to validation.h functions const

TheBlueMatt committed:
Move LoadChainTip into CChainState

TheBlueMatt committed:
Make all CBlockIndex*es in validation.h const

TheBlueMatt committed:
mapBlockIndex only in validation.cpp, and const outside CChainState

TheBlueMatt committed:
Remove now-useless const_cast in qt/clientmodel

TheBlueMatt committed:
scripted-diff: Use nonCosntChainActive in miner_tests

-BEGIN VERIFY SCRIPT-
sed -i "s/chainActive\./nonConstChainActive./" src/test/miner_tests.cpp
-END VERIFY SCRIPT-

TheBlueMatt committed:
Make VerifyDB a CChainState member instead of a loose class

TheBlueMatt committed:
Make chainActive const outside of CChainState

Empact commented on src/chain.cpp:
```diff
@@ -59,12 +59,17 @@ const CBlockIndex *CChain::FindFork(const CBlockIndex *pindex) const {
+    std::vector<CBlockIndex*>::const_iterator lower = std::lower_bound(vChain.begin(), vChain.end(), nTime,
+        [](CBlockIndex* pBlock, const int64_t& time) -> bool { return pBlock->GetBlockTimeMax() < time; });
+    return (lower == vChain.end() ? nullptr : *lower);
+}
```
Compiles without this definition if you fix the skiplist_tests use to be const as well.

TheBlueMatt replied:
Done.

TheBlueMatt committed:
scripted-diff: constify CBlockIndex* outside of validation

-BEGIN VERIFY SCRIPT-
sed -i "s/\(const \)*CBlockIndex\( \)*\*\( \)*/const CBlockIndex\* /g" src/checkpoints.h src/checkpoints.cpp src/rest.cpp src/rpc/*.cpp src/wallet/*.h src/wallet/*.cpp src/init.cpp src/qt/clientmodel.cpp src/miner.cpp
-END VERIFY SCRIPT-

TheBlueMatt committed:
Constify CBlockIndex*s in skiplist_tests where possible

promag commented on src/validation.cpp:
```diff
@@ -3849,26 +3849,30 @@ bool LoadChainTip(const CChainParams& chainparams)
         LogPrintf("%s: Connecting genesis block...\n", __func__);
         CValidationState state;
-        if (!ActivateBestChain(state, chainparams)) {
+        if (!ActivateBestChain(state, chainparams, std::shared_ptr<const CBlock>())) {
```
In commit "Move LoadChainTip into CChainState",

Use `MakeShared()`?

TheBlueMatt replied:
shared_ptr() creates a pointer-to-null, not a default-constructed CBlock.

jimpo replied:
Any particular reason to pass it explicitly?

promag replied:
Indeed, this change could be reverted (it's the default value).

promag commented on src/validation.cpp:
```diff
@@ -191,6 +190,7 @@ class CChainState {
     CBlockIndex* FindMostWorkChain();
     bool ReceivedBlockTransactions(const CBlock &block, CValidationState& state, CBlockIndex *pindexNew, const CDiskBlockPos& pos, const Consensus::Params& consensusParams);
 
+    void PruneBlockIndexCandidates();
```
In commit "Move LoadChainTip into CChainState",

Nit, revert this unrelated change?

TheBlueMatt replied:
How is this unrelated? Because of the change PruneBlockIndexCandidates (an internal helper function) is no longer called outside of CChainState, so it should be made private.

promag replied:
Oh missed the visibility change.

promag commented on src/test/miner_tests.cpp:
```diff
@@ -39,6 +39,8 @@ class HasReason {
 
 static CFeeRate blockMinFeeRate = CFeeRate(DEFAULT_BLOCK_MIN_TX_FEE);
 
+static CChain& nonConstChainActive = const_cast<CChain&>(chainActive);
```
In commit "Add a const_cast'ed reference to chainActive in miner_tests",

Why add an unused variable? Could add just when it's needed?

Answer: Because later it's used in "scripted-diff: Use nonCosntChainActive in miner_tests" commit.

promag reviewed (commented):
Needs rebase. Looks good.

sipa commented:
utACK 4576f1de6725bab583797d081421b992872fd616

Nit: typo in commit name `nonCosntChainActive`.

jimpo commented on src/validation.cpp:
```diff
@@ -177,6 +178,13 @@ class CChainState {
     void UnloadBlockIndex();
 
+    ~CChainState() {
+        BlockMap::iterator it1 = mapBlockIndex.begin();
```
nit: This can be simplified using the foreach loop in `UnloadBlockIndex`. And the explicit clear at the end is unnecessary.

```c++
for (BlockMap::value_type& entry : mapBlockIndex) delete entry.second;
```

promag replied:
Commit "mapBlockIndex only in validation.cpp, and const outside CChainState"

> And the explicit clear at the end is unnecessary.

@jimpo why? Otherwise it would contain dangling pointers. `mapBlockIndex` isn't a `CChainState` member.

jimpo reviewed (commented):
Looks good. Needs rebase.

jnewbery commented:
PR would be more searchable with a descriptive title :)

Empact commented:
nit: use `?` in regex for 0-1 copies, rather than `*` (0-n copies)

sipa commented:
Needs rebase due to #10244.

promag commented on src/validation.cpp:
```diff
@@ -166,18 +169,24 @@ class CChainState {
     void UnloadBlockIndex();
 
+    ~CChainState() {
+        BlockMap::iterator it1 = mapBlockIndex.begin();
```
Commit "mapBlockIndex only in validation.cpp, and const outside CChainState"

Should `AssertLockHeld(cs_main)`?

promag commented on src/qt/clientmodel.cpp:
```diff
@@ -137,7 +137,7 @@ size_t ClientModel::getMempoolDynamicUsage() const
 double ClientModel::getVerificationProgress(const CBlockIndex* tipIn) const
 {
-    const CBlockIndex* tip = const_cast<const CBlockIndex* >(tipIn);
+    const CBlockIndex* tip = tipIn;
```
Commit "Remove now-useless const_cast in qt/clientmodel"

Could remove this line and rename argument to `tip`.

promag reviewed (commented):
Typo in commit "scripted-diff: Use nonCosntChainActive in miner_tests"
